core-bundle/src/Routing/ResponseContext/ResponseContext.php

```diff
@@ -12,31 +12,79 @@
 
 namespace Contao\CoreBundle\Routing\ResponseContext;
 
-use Symfony\Component\HttpFoundation\Response;
-use Symfony\Component\HttpFoundation\ResponseHeaderBag;
+use Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException;
 
-class ResponseContext implements ResponseContextInterface
+final class ResponseContext
 {
+    public const REQUEST_ATTRIBUTE_NAME = '_contao_response_context';
+
+    /**
+     * @var array<string,\Closure>
+     */
+    private $services;
+
     /**
      * @var PartialResponseHeaderBag|null
      */
     private $headerBag;
 
-    public function getHeaderBag(): ResponseHeaderBag
+    public function add(object $service): self
     {
-        if (null === $this->headerBag) {
-            $this->headerBag = new PartialResponseHeaderBag();
+        return $this->addLazy(\get_class($service), static function () use ($service) { return $service; });
+    }
+
+    public function addLazy(string $classname, \Closure $factory)
+    {
+        $this->services[$classname] = $factory;
+
+        $ref = new \ReflectionClass($classname);
+
+        // Automatically add aliases for all interfaces and parents (last one added automatically wins by overriding here)
+        foreach ($ref->getInterfaceNames() as $interfaceName) {
+            $this->services[$interfaceName] = $factory;
         }
 
-        return $this->headerBag;
+        while ($ref = $ref->getParentClass()) {
+            $this->services[$ref->getName()] = $factory;
+        }
```

Comment:
Unfortunately, this isn't going to work, since accessing the same service through multiple classes would call the factory for each of them I think

Reply:
What doesn't work? The tests work?

Reply:
Since you're saving the factory method for multiple service keys, accessing each of the keys will separately call it's factory.

**Example:**
```
$context->addLazy('Foobar extends FooInterface', function () { echo 'test' });

$context->get('Foobar');
$context->get('FooInterface');
```

this would `echo 'test'` twice

Reply:
I don't get the problem, sorry.

Reply:
Maybe you can PR a failing test?

Reply:
Here you go: https://github.com/Toflar/contao/pull/4
Also, please don't resolve an open conversation…

Reply:
Fixed in https://github.com/contao/contao/pull/3066/commits/1d29c95c52158ba7e31347898071e7b5c125bcfc.